did_parser/src/parsed_did_url.rs

```diff
@@ -0,0 +1,200 @@
+        }
+
+        Ok(DidUrl {
+            did_url,
+            did,
+            method,
+            id,
+            path,
+            queries,
+            fragment,
+            params,
+        })
+    }
+
+    pub fn did_url(&self) -> &str {
+        self.did_url.as_ref()
+    }
+
+    pub fn did(&self) -> Option<&str> {
+        self.did.clone().map(|range| self.did_url[range].as_ref())
+    }
+
+    pub fn method(&self) -> Option<&str> {
+        self.method
+            .clone()
+            .map(|range| self.did_url[range].as_ref())
+    }
+
+    pub fn id(&self) -> Option<&str> {
+        self.id.clone().map(|range| self.did_url[range].as_ref())
+    }
+
+    pub fn path(&self) -> Option<&str> {
+        self.path.as_ref().map(|path| &self.did_url[path.clone()])
+    }
+
+    pub fn queries(&self) -> HashMap<String, String> {
+        self.queries
+            .iter()
+            .map(|(k, v)| {
+                (
+                    self.did_url[k.clone()].to_string(),
+                    self.did_url[v.clone()].to_string(),
+                )
+            })
+            .collect()
+    }
+
+    pub fn fragment(&self) -> Option<&str> {
+        self.fragment
+            .as_ref()
+            .map(|fragment| &self.did_url[fragment.clone()])
+    }
+
+    pub fn params(&self) -> HashMap<String, String> {
+        self.params
+            .iter()
+            .map(|(k, v)| {
+                (
+                    self.did_url[k.clone()].to_string(),
+                    self.did_url[v.clone()].to_string(),
+                )
+            })
+            .collect()
+    }
+}
+
+impl TryFrom<String> for DidUrl {
+    type Error = ParseError;
+
+    fn try_from(did_url: String) -> Result<Self, Self::Error> {
+        Self::parse(did_url)
+    }
+}
+
+impl Display for DidUrl {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.did_url)
+    }
+}
+
+impl Serialize for DidUrl {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&self.did_url)
+    }
+}
+
+impl<'de> Deserialize<'de> for DidUrl {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let did_url = String::deserialize(deserializer)?;
+        DidUrl::parse(did_url).map_err(serde::de::Error::custom)
+    }
+}
+
+impl TryFrom<&DidUrl> for Did {
+    type Error = ParseError;
+
+    fn try_from(did_url: &DidUrl) -> Result<Self, Self::Error> {
+        if let (Some(did), Some(method), Some(id)) = (did_url.did(), &did_url.method, &did_url.id) {
+            Ok(Did::from_parts(
+                did.to_owned(),
+                method.to_owned(),
+                id.to_owned(),
+            ))
+        } else {
+            Err(ParseError::InvalidInput(
+                "Unable to construct a DID from relative DID URL",
+            ))
+        }
+    }
+}
```

Comment:
Why from a `&DidUrl` and not a `DidUrl`?

Reply:
Is there a need to own it for the conversion? It's used here:

```
    async fn dereference(
        &self,
        did_url: &DidUrl,
        _options: &DidDereferencingOptions,
    ) -> Result<DidDereferencingOutput<Self::Output>, GenericError> {
        let resolution_output = self
            .resolve(&did_url.try_into()?, &DidResolutionOptions::default())
            .await?;

        dereference_did_document(&resolution_output, did_url).map_err(|err| err.into())
    }
```

Reply:
It's more idiomatic. This limits the consumer API to only convert from a reference. This method can easily clone just as the consumer code can. The compiler is surely smart enough not to clone the other parts of `DidUrl` that are not used here in release mode.

Reply:
Maybe we're looking at this the wrong way. What if the `Did` and `DidUrl` types would implement a common trait exposing `did`, `method` and `id`? Then the `resolve` method could take a generic implementing the trait and the conversion wouldn't even be necessary. 

What do you think?

Reply:
Implementing only owned version limits the consumer API as it must own `DidUrl`. There is no need to clone here as well as there is no need to rely on compiler optimizations.

Reply:
Okay, this is your PR, you call the shots. I'm at least glad the `TryFrom` impl has changed. While I don't understand `asm`, parsing the `String` outputs 152 lines of code while the new impl outputs 37 lines. I find that to be huge.